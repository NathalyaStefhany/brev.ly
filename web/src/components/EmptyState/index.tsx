import React, { ComponentProps, ReactElement } from 'react';

type EmptyStateProps = {
  icon: ReactElement;
  description: string;
} & ComponentProps<'div'>;

export const EmptyState: React.FC<EmptyStateProps> = ({
  icon,
  description,
  ...props
}) => {
  return (
    <div
      className="flex flex-col items-center justify-center gap-6 mt-16 mb-12"
      {...props}
    >
      {icon}

      <p className="text-xs text-gray-500 uppercase text-center">
        {description}
      </p>
    </div>
  );
};
