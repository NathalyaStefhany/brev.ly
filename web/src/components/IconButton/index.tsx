import React, { ComponentProps, ComponentType } from 'react';

import { IconProps } from '@phosphor-icons/react';

type IconButtonProps = ComponentProps<'button'> & {
  icon: ComponentType<IconProps>;
};

export const IconButton: React.FC<IconButtonProps> = ({
  icon: Icon,
  ...props
}) => {
  return (
    <button
      type="button"
      className="cursor-pointer transition-all duration-300 bg-gray-200 border border-gray-200 rounded-sm p-3.5 hover:not-disabled:border-blue-base disabled:opacity-50 disabled:cursor-not-allowed"
      {...props}
    >
      <Icon size="1rem" color="var(--color-gray-600)" />
    </button>
  );
};
