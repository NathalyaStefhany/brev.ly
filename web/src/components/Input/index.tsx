import { Warning } from '@phosphor-icons/react';
import React, { ComponentProps } from 'react';

import { tv } from 'tailwind-variants';

const labelVariants = tv({
  base: 'text-xs text-gray-500 uppercase focus:font-bold peer-focus:font-bold',
  variants: {
    intent: {
      default: 'peer-focus:text-blue-base',
      error: 'peer-focus:text-danger',
    },
  },
  defaultVariants: {
    intent: 'default',
  },
});

const inputVariants = tv({
  base: 'peer text-md text-gray-600 font-normal border border-gray-300 caret-blue-base rounded-lg px-8 py-7 placeholder:text-gray-400',
  variants: {
    intent: {
      default: 'focus:outline-blue-base',
      error: 'focus:outline-danger',
    },
  },
  defaultVariants: {
    intent: 'default',
  },
});

type InputProps = ComponentProps<'input'> & {
  id: string;
  label: string;
  error?: string;
};

export const Input: React.FC<InputProps> = ({ id, label, error, ...props }) => {
  const intent = error ? 'error' : 'default';

  return (
    <div className="max-w-176 w-full flex flex-col-reverse gap-4">
      {error && (
        <div className="flex flex-row items-center gap-4">
          <Warning size="1rem" color="var(--color-danger)" />

          <span
            id={`${id}-error`}
            className="text-sm text-gray-500"
            data-testid={`${id}-error`}
          >
            {error}
          </span>
        </div>
      )}

      <input
        id={id}
        type="text"
        aria-invalid={!!error}
        aria-describedby={`${id}-error`}
        className={inputVariants({ intent })}
        {...props}
      />

      <label htmlFor={id} className={labelVariants({ intent })}>
        {label}
      </label>
    </div>
  );
};
