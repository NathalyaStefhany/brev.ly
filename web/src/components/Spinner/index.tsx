import React from 'react';

export const Spinner: React.FC = () => {
  return (
    <div className="relative w-12 h-12">
      {[...Array(10)].map((_, i) => (
        <div
          // eslint-disable-next-line react/no-array-index-key
          key={i}
          className="absolute top-4.5 left-5.5 w-1 h-3 bg-blue-base rounded origin-center animate-fade"
          style={{
            transform: `rotate(${i * 36}deg) translateY(-150%)`,
            opacity: i / 10,
            animationDelay: `${i * 0.098}s`,
          }}
        />
      ))}
    </div>
  );
};
