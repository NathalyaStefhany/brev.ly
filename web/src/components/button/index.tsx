import React, { ComponentProps, ComponentType } from 'react';

import { tv, VariantProps } from 'tailwind-variants';
import { IconProps } from '@phosphor-icons/react';

const buttonVariants = tv({
  base: 'flex flex-row gap-3 items-center justify-center cursor-pointer transition-all duration-300 disabled:cursor-not-allowed',
  variants: {
    variant: {
      primary:
        'max-w-176 w-full bg-blue-base text-md text-white rounded-lg px-10 py-7.5 hover:not-disabled:bg-blue-dark disabled:opacity-50',
      secondary:
        'bg-gray-200 border border-gray-200 text-sm font-semibold text-gray-500 rounded-sm p-3.5 hover:not-disabled:border-blue-base disabled:opacity-50',
    },
  },
  defaultVariants: {
    variant: 'primary',
  },
});

type ButtonProps = ComponentProps<'button'> &
  VariantProps<typeof buttonVariants> & {
    icon?: ComponentType<IconProps>;
  };

export const Button: React.FC<ButtonProps> = ({
  variant,
  icon: Icon,
  children,
  ...props
}) => {
  return (
    <button type="button" className={buttonVariants({ variant })} {...props}>
      {Icon && (
        <Icon
          size="1rem"
          color="var(--color-gray-600)"
          data-testid="button-icon"
        />
      )}

      {children}
    </button>
  );
};
