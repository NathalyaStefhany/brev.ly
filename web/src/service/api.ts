import { env } from '@/env';
import axios from 'axios';

const baseURL = env.VITE_BACKEND_URL;

export const api = axios.create({
  baseURL,
});
