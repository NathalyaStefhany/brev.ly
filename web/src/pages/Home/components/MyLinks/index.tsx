import { Button } from '@/components/Button';
import { DownloadSimple, Link } from '@phosphor-icons/react';
import React from 'react';

export const MyLinks: React.FC = () => {
  const listOfLinksIsEmpty = true;

  return (
    <div className="max-w-190 w-full md:max-w-290 md:min-w-190 h-fit flex flex-col flex-1 gap-8 md:gap-10 bg-gray-100 rounded-lg p-12 md:p-16">
      <div className="flex flex-row items-center justify-between">
        <h2 className="text-lg text-gray-600">Meus links</h2>

        <Button variant="secondary" icon={DownloadSimple} disabled>
          Baixar CSV
        </Button>
      </div>

      <div className="border-t border-gray-200">
        {listOfLinksIsEmpty && (
          <div className="flex flex-col items-center justify-center gap-6 mt-8 mb-12">
            <Link size="2rem" color="var(--color-gray-400)" />

            <p className="text-xs text-gray-500 uppercase text-center">
              Ainda n√£o existem links cadastrados
            </p>
          </div>
        )}
      </div>
    </div>
  );
};
