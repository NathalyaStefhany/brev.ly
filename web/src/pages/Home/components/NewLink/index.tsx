import React, { useState } from 'react';

import { useForm } from 'react-hook-form';
import * as Toast from '@radix-ui/react-toast';
import { zodResolver } from '@hookform/resolvers/zod';
import { z } from 'zod';
import { Input } from '@/components/Input';
import { Button } from '@/components/Button';
import { api } from '@/service/api';
import { WarningCircle, X } from '@phosphor-icons/react';
import { AxiosError } from 'axios';
import { queryClient } from '@/service/queryClient';

const originalLinkErrorMessage = 'Informe uma url válida.';
const shortenedLinkErrorMessage =
  'Informe uma url minúscula e sem espaço/caracter especial.';

const newLinkSchema = z
  .object({
    originalLink: z
      .string()
      .min(1, originalLinkErrorMessage)
      .refine((value) => {
        try {
          new URL(value);

          return true;
        } catch {
          return /^(www\.)[\w-]+\.[a-z]{2,}(\.[a-z]{2,})?(\/[^\s]*)?$/i.test(
            value,
          );
        }
      }, originalLinkErrorMessage),
    shortenedLink: z
      .string()
      .min(1, shortenedLinkErrorMessage)
      .regex(/^[a-z0-9-]+$/, shortenedLinkErrorMessage),
  })
  .required();

type NewLinkData = z.infer<typeof newLinkSchema>;

type ErrorData = {
  message?: string;
};

export const NewLink: React.FC = () => {
  const [openToastError, setOpenToastError] = useState(false);
  const [errorCreatingNewLink, setErrorCreatingNewLink] = useState({
    title: '',
    description: '',
  });

  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
    reset,
  } = useForm({
    resolver: zodResolver(newLinkSchema),
  });

  const handleSaveLink = async (data: NewLinkData): Promise<void> => {
    try {
      await api.post('/shortened-links', data);

      queryClient.refetchQueries({ queryKey: ['links list'], exact: true });

      reset();
    } catch (error) {
      const err = error as AxiosError;

      let description = 'Por favor, tente novamente mais tarde.';

      const errorData = err.response?.data as ErrorData;

      if (
        err.status === 400 &&
        errorData?.message === 'Shortened link already exists'
      ) {
        description = 'Esse link encurtado já existe.';
      }

      setErrorCreatingNewLink({
        title: 'Erro no cadastro',
        description,
      });

      setOpenToastError(true);
    }
  };

  return (
    <>
      <form
        onSubmit={handleSubmit(handleSaveLink)}
        className="max-w-190 w-full md:min-w-145 flex flex-col flex-1 gap-10 md:gap-12 bg-gray-100 rounded-lg p-12 md:p-16"
      >
        <h2 className="text-lg text-gray-600">Novo link</h2>

        <div className="flex flex-col gap-8">
          <Input
            id="input-original-link"
            label="Link Original"
            placeholder="www.exemplo.com.br"
            error={errors.originalLink?.message}
            data-testid="input-original-link"
            {...register('originalLink')}
          />

          <Input
            id="input-shortened-link"
            label="Link Encurtado"
            fixedPlaceholder="brev.ly/"
            error={errors.shortenedLink?.message}
            data-testid="input-shortened-link"
            {...register('shortenedLink')}
          />
        </div>

        <Button
          type="submit"
          disabled={isSubmitting}
          data-testid="button-save-link"
        >
          {isSubmitting ? 'Salvando...' : 'Salvar link'}
        </Button>
      </form>

      <Toast.Provider swipeDirection="right">
        <Toast.Root
          open={openToastError}
          onOpenChange={setOpenToastError}
          duration={5000}
          className="w-150 flex flex-row gap-6 items-start justify-start bg-[#f1d4da] p-8 rounded-lg shadow-lg shadow-gray-300"
          data-testid="toast-creation-error"
        >
          <WarningCircle
            color="var(--color-danger)"
            size="1.25rem"
            weight="fill"
          />

          <div className="flex-1">
            <Toast.Title className="text-md text-danger">
              {errorCreatingNewLink.title}
            </Toast.Title>

            <Toast.Description className="text-sm text-danger mt-2">
              {errorCreatingNewLink.description}
            </Toast.Description>
          </div>

          <Toast.Action
            altText="Botão para fechar a caixa de erro"
            className="cursor-pointer text-danger"
          >
            <X />
          </Toast.Action>
        </Toast.Root>

        <Toast.Viewport className="fixed bottom-6 left-1/2 transform -translate-x-1/2 md:bottom-12 md:right-12 md:left-auto md:transform-none md:translate-x-0" />
      </Toast.Provider>
    </>
  );
};
