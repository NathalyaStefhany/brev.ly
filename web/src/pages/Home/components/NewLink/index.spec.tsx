import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import AxiosMock from 'axios-mock-adapter';
import { NewLink } from '@/pages/Home/components/NewLink';
import { api } from '@/service/api';
import { queryClient } from '@/service/queryClient';

const apiMock = new AxiosMock(api);

describe('New Link tests', () => {
  beforeEach(() => {
    apiMock.reset();
  });

  it('should render correctly', async () => {
    let container: HTMLElement = document.createElement('div');

    // eslint-disable-next-line @typescript-eslint/require-await
    await act(async () => {
      container = render(<NewLink />).container;
    });

    expect(container).toMatchSnapshot();
  });

  it('should display input errors when try to save link but the fields are empty', async () => {
    render(<NewLink />);

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(screen.getByTestId('input-original-link-error')).toHaveTextContent(
        'Informe uma url válida.',
      );
    });

    expect(screen.getByTestId('input-shortened-link-error')).toHaveTextContent(
      'Informe uma url minúscula e sem espaço/caracter especial.',
    );
  });

  it('should display original link input error when value is invalid (example.com)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'example.com' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'example.com',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should display original link input error when value is invalid (https://)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'https://' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue('https://');

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should display original link input error when value is invalid (https://example@)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'https://example@' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'https://example@',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should display original link input error when value is invalid (www.example)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.example' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'www.example',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should display original link input error when value is invalid (www.example.com.br.test)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.example.com.br.test' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'www.example.com.br.test',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should display original link input error when value is invalid (www.example.)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.example.' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'www.example.',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should not display original link input error when value is valid (https://example.com)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'https://example.com' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'https://example.com',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });

    expect(
      screen.queryByTestId('input-original-link-error'),
    ).not.toBeInTheDocument();
  });

  it('should not display original link input error when value is valid (http://example.com)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'http://example.com' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'http://example.com',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });

    expect(
      screen.queryByTestId('input-original-link-error'),
    ).not.toBeInTheDocument();
  });

  it('should not display original link input error when value is valid (https://www.example.com)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'https://www.example.com' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'https://www.example.com',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });

    expect(
      screen.queryByTestId('input-original-link-error'),
    ).not.toBeInTheDocument();
  });

  it('should not display original link input error when value is valid (https://example.com.br)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'https://example.com.br' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'https://example.com.br',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });

    expect(
      screen.queryByTestId('input-original-link-error'),
    ).not.toBeInTheDocument();
  });

  it('should not display original link input error when value is valid (www.example.com)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.example.com' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'www.example.com',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });

    expect(
      screen.queryByTestId('input-original-link-error'),
    ).not.toBeInTheDocument();
  });

  it('should not display original link input error when value is valid (www.example.com.br)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.example.com.br' },
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'www.example.com.br',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });

    expect(
      screen.queryByTestId('input-original-link-error'),
    ).not.toBeInTheDocument();
  });

  it('should display shortened link input error when value is invalid (EXAMPLE)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'EXAMPLE' },
    });

    expect(screen.getByTestId('input-shortened-link')).toHaveValue('EXAMPLE');

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should display shortened link input error when value is invalid (example.)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'example.' },
    });

    expect(screen.getByTestId('input-shortened-link')).toHaveValue('example.');

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should display shortened link input error when value is invalid (example )', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'example ' },
    });

    expect(screen.getByTestId('input-shortened-link')).toHaveValue('example ');

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should not display shortened link input error when value is valid (example)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'example' },
    });

    expect(screen.getByTestId('input-shortened-link')).toHaveValue('example');

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });

    expect(
      screen.queryByTestId('input-shortened-link-error'),
    ).not.toBeInTheDocument();
  });

  it('should not display shortened link input error when value is valid (example-1)', async () => {
    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'example-1' },
    });

    expect(screen.getByTestId('input-shortened-link')).toHaveValue('example-1');

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });

    expect(
      screen.queryByTestId('input-shortened-link-error'),
    ).not.toBeInTheDocument();
  });

  it('should disappear with the original input error when entering a valid value', async () => {
    render(<NewLink />);

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.example-1.com' },
    });

    await waitFor(() => {
      expect(
        screen.queryByTestId('input-original-link-error'),
      ).not.toBeInTheDocument();
    });

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www' },
    });

    await waitFor(() => {
      expect(
        screen.getByTestId('input-original-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should disappear with the shortened input error when entering a valid value', async () => {
    render(<NewLink />);

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'example' },
    });

    await waitFor(() => {
      expect(
        screen.queryByTestId('input-shortened-link-error'),
      ).not.toBeInTheDocument();
    });

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'exâmple' },
    });

    await waitFor(() => {
      expect(
        screen.getByTestId('input-shortened-link-error'),
      ).toBeInTheDocument();
    });
  });

  it('should be able to create a new link', async () => {
    const newLink = {
      originalLink: 'www.test.com',
      shortenedLink: 'test',
    };

    apiMock.onPost('/shortened-links', newLink).reply(200);

    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.test.com' },
    });

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'test' },
    });

    expect(screen.getByTestId('button-save-link')).toHaveTextContent(
      'Salvar link',
    );

    fireEvent.click(screen.getByTestId('button-save-link'));

    expect(screen.getByTestId('button-save-link')).toBeDisabled();
    expect(screen.getByTestId('button-save-link')).toHaveTextContent(
      'Salvando...',
    );

    await waitFor(() => {
      expect(
        apiMock.history.post.some(
          ({ url, data }) =>
            url === '/shortened-links' && data === JSON.stringify(newLink),
        ),
      ).toBeTruthy();
    });

    expect(
      screen.queryByTestId('toast-creation-error'),
    ).not.toBeInTheDocument();
  });

  it('should display error when try to create a new link with a shortened link that already exists', async () => {
    const newLink = {
      originalLink: 'www.test.com',
      shortenedLink: 'test',
    };

    apiMock
      .onPost('/shortened-links', newLink)
      .reply(400, { message: 'Shortened link already exists' });

    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.test.com' },
    });

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'test' },
    });

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        apiMock.history.post.some(
          ({ url, data }) =>
            url === '/shortened-links' && data === JSON.stringify(newLink),
        ),
      ).toBeTruthy();
    });

    expect(screen.getByTestId('toast-creation-error')).toBeInTheDocument();
    expect(screen.getByTestId('toast-creation-error')).toHaveTextContent(
      'Esse link encurtado já existe.',
    );
  });

  it('should display error when creation fails', async () => {
    const newLink = {
      originalLink: 'www.test.com',
      shortenedLink: 'test',
    };

    apiMock.onPost('/shortened-links', newLink).reply(400);

    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.test.com' },
    });

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'test' },
    });

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        apiMock.history.post.some(
          ({ url, data }) =>
            url === '/shortened-links' && data === JSON.stringify(newLink),
        ),
      ).toBeTruthy();
    });

    expect(screen.getByTestId('toast-creation-error')).toBeInTheDocument();
    expect(screen.getByTestId('toast-creation-error')).toHaveTextContent(
      'Por favor, tente novamente mais tarde.',
    );
  });

  it('should reset input values when create a new link', async () => {
    const newLink = {
      originalLink: 'www.test.com',
      shortenedLink: 'test',
    };

    apiMock.onPost('/shortened-links', newLink).reply(200);

    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.test.com' },
    });

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'test' },
    });

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        apiMock.history.post.some(
          ({ url, data }) =>
            url === '/shortened-links' && data === JSON.stringify(newLink),
        ),
      ).toBeTruthy();
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue('');
    expect(screen.getByTestId('input-shortened-link')).toHaveValue('');
  });

  it('should not reset input values when creation fails', async () => {
    const newLink = {
      originalLink: 'www.test.com',
      shortenedLink: 'test',
    };

    apiMock.onPost('/shortened-links', newLink).reply(400);

    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.test.com' },
    });

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'test' },
    });

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        apiMock.history.post.some(
          ({ url, data }) =>
            url === '/shortened-links' && data === JSON.stringify(newLink),
        ),
      ).toBeTruthy();
    });

    expect(screen.getByTestId('input-original-link')).toHaveValue(
      'www.test.com',
    );
    expect(screen.getByTestId('input-shortened-link')).toHaveValue('test');
  });

  it('should refetch links list query when create a new link', async () => {
    const newLink = {
      originalLink: 'www.test.com',
      shortenedLink: 'test',
    };

    apiMock.onPost('/shortened-links', newLink).reply(200);

    queryClient.refetchQueries = vi.fn();

    render(<NewLink />);

    fireEvent.change(screen.getByTestId('input-original-link'), {
      target: { value: 'www.test.com' },
    });

    fireEvent.change(screen.getByTestId('input-shortened-link'), {
      target: { value: 'test' },
    });

    fireEvent.click(screen.getByTestId('button-save-link'));

    await waitFor(() => {
      expect(
        apiMock.history.post.some(
          ({ url, data }) =>
            url === '/shortened-links' && data === JSON.stringify(newLink),
        ),
      ).toBeTruthy();
    });

    expect(queryClient.refetchQueries).toBeCalledWith({
      queryKey: ['links list'],
      exact: true,
    });
  });
});
