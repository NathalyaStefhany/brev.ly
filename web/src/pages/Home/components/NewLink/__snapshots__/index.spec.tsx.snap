// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`New Link tests > should render correctly 1`] = `
<div>
  <form
    class="max-w-190 w-full md:min-w-145 flex flex-col flex-1 gap-10 md:gap-12 bg-gray-100 rounded-lg p-12 md:p-16"
  >
    <h2
      class="text-lg text-gray-600"
    >
      Novo link
    </h2>
    <div
      class="flex flex-col gap-8"
    >
      <div
        class="relative max-w-176 w-full flex flex-col-reverse gap-4"
      >
        <input
          aria-describedby="input-original-link-error"
          aria-invalid="false"
          class="peer z-1 text-md text-gray-600 font-normal border border-gray-300 caret-blue-base rounded-lg py-7 placeholder:text-gray-400 focus:outline-blue-base"
          data-testid="input-original-link"
          id="input-original-link"
          name="originalLink"
          placeholder="www.exemplo.com.br"
          style="padding-left: calc(1rem); padding-right: 1rem;"
          type="text"
        />
        <label
          class="text-xs text-gray-500 uppercase focus:font-bold peer-focus:font-bold peer-focus:text-blue-base"
          for="input-original-link"
        >
          Link Original
        </label>
      </div>
      <div
        class="relative max-w-176 w-full flex flex-col-reverse gap-4"
      >
        <input
          aria-describedby="input-shortened-link-error"
          aria-invalid="false"
          class="peer z-1 text-md text-gray-600 font-normal border border-gray-300 caret-blue-base rounded-lg py-7 placeholder:text-gray-400 focus:outline-blue-base"
          data-testid="input-shortened-link"
          id="input-shortened-link"
          name="shortenedLink"
          style="padding-left: calc(1.5rem); padding-right: 1rem;"
          type="text"
        />
        <span
          class="absolute top-18.5 left-8.5 text-md text-gray-400 font-normal"
          data-testid="input-shortened-link-fixed-placeholder"
        >
          brev.ly/
        </span>
        <label
          class="text-xs text-gray-500 uppercase focus:font-bold peer-focus:font-bold peer-focus:text-blue-base"
          for="input-shortened-link"
        >
          Link Encurtado
        </label>
      </div>
    </div>
    <button
      class="flex flex-row gap-3 items-center justify-center cursor-pointer transition-all duration-300 disabled:cursor-not-allowed max-w-176 w-full bg-blue-base text-md text-white rounded-lg px-10 py-7.5 hover:not-disabled:bg-blue-dark disabled:opacity-50"
      data-testid="button-save-link"
      type="submit"
    >
      Salvar link
    </button>
  </form>
</div>
`;
